### PR TITLE
Allow casting to boolean in Serverless variables

### LINKED ISSUE #6782: The always restApi logs enable when using an SSM parameter in the configuration.
# Bug Report

## Description

1. What did you do?

I enable logs depending on the stage using SSM parameters.

1. What happened?

The API Gateway logs enable for every stage. The SSM values for `dev` and `test` stages were `false`, and `staging` and `prod` were `true`.

1. What should've happened?

API Gateway logs should have been disabled for `dev` and `test`, and enabled for `staging` and `prod`.

1. What's the content of your `serverless.yml` file?

```yaml
provider:
  logs:
    restApi:
      accessLogging:     ${ssm:/${self:provider.env}/API_GW_LOGS}
      executionLogging:  ${ssm:/${self:provider.env}/API_GW_LOGS}
      fullExecutionData: ${ssm:/${self:provider.env}/API_GW_LOGS}
```

1. What's the output you get when you use the `SLS_DEBUG=*` environment variable (e.g. `SLS_DEBUG=* serverless deploy`)

Not captured.

Similar or dependent issues:

- #12345

### PR DESCRIPTION
**Pull request by miguel-a-calles-mba**
## What did you implement

Variables can now cast values to booleans. For example, if a parameter expects a boolean, but a string is passed, it will convert a `"true"` to `true`, a `"false"` to `false`, and `"anything"` to `true`.

Closes #6782

## How can we verify it

Reference an SSM value and cast it in the API Gateway logging and X-ray tracing provider sections.

```yaml
  tracing:
    apiGateway: ${strToBool(${ssm:API_GW_DEBUG_ENABLED})}

  logs:
    restApi:
      accessLogging:     ${strToBool(${ssm:API_GW_DEBUG_ENABLED})}
      executionLogging:  ${strToBool(${ssm:API_GW_DEBUG_ENABLED})}
      fullExecutionData: ${strToBool(${ssm:API_GW_DEBUG_ENABLED})}

```

Set `API_GW_DEBUG_ENABLED` to `true` in SSM, deploy and see the API Gateway settings turn on. Update `API_GW_DEBUG_ENABLED` to `false` in SSM, deploy and see the API Gateway settings turn off.

## Todos

<details>
<summary>Useful Scripts</summary>

- `npm run test-ci` --> Run all validation checks on proposed changes
- `npm run lint-updated` --> Lint all the updated files
- `npm run lint:fix` --> Automatically fix lint problems (if possible)
- `npm run prettier-check-updated` --> Check if updated files adhere to Prettier config
- `npm run prettify-updated` --> Prettify all the updated files

</details>

- [X] Write and run all tests
- [x] Write documentation
- [X] Enable "Allow edits from maintainers" for this PR
- [X] Update the messages below

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO

### CODE REVIEW

**Comment by miguel-a-calles-mba** [comment]
It works really nicely in the API Gateway tracing and logging options.

File: lib/classes/Variables.js [hunk 1]
```diff
@@ -55,6 +55,7 @@ class Variables {
     this.cfRefSyntax = RegExp(/^(?:\${)?cf(\.[a-zA-Z0-9-]+)?:/g);
     this.s3RefSyntax = RegExp(/^(?:\${)?s3:(.+?)\/(.+)$/);
     this.ssmRefSyntax = RegExp(/^(?:\${)?ssm:([a-zA-Z0-9_.\-/]+)[~]?(true|false|split)?/);
+    this.toBoolRefSyntax = RegExp(/^(?:\${)?toBool:([a-zA-Z0-9_.\-/]+)/);
```
Comment (miguel-a-calles-mba) [inline lib/classes/Variables.js]: I can change this to `${toBool(value)}` if that is preferred over `${toBool:value}`.

File: docs/providers/aws/guide/variables.md [hunk 2]
```diff
@@ -657,3 +660,15 @@ Resources:
...
+## Casting string variables values to boolean
+
+In some cases, a parameter expect a `true` or `false` value. If you are using a variable to define the value, it may return as a string (e.g. when using SSM variables) and thus return a `"true"` or `"false"` value.
+
+To ensure a boolean value is returned, cast the variable value. For example:
+
+```yml
+provider:
+  tracing:
+    apiGateway: ${toBool:${ssm:API_GW_DEBUG_ENABLED}}
```
Comment (miguel-a-calles-mba) [inline docs/providers/aws/guide/variables.md]: I can change this to `${toBool(value)}` if that is preferred over `${toBool:value}`.

Comment (medikoo) [reply docs/providers/aws/guide/variables.md]: I think it's important to name it as `strToBool` or `stringToBool`.. as otherwise `toBool` suggests language native boolean convertion (which can be assumed as an alias for `Boolean(value)`, when it's not what it does.

Here we're doing something custom (technically it's _boolean string to boolean_ coercion, and not _value to boolean_). So even more accurate name would be `boolStrToBool` but that might be too verbose.

@pmuens thoughts?

Comment (miguel-a-calles-mba) [reply docs/providers/aws/guide/variables.md]: I see your point. Using `to` in the name will imply casting. What if we use `as` in the name to imply translation. For example, what if we use `strAsBool` or `readStringAsBoolean`?

Comment (medikoo) [reply docs/providers/aws/guide/variables.md]: > Using to in the name will imply casting

It is about casting, so `to` sounds fine to me.. Still it should not imply that we simply cast to _boolean_ per language rules (as that's already done internally, no need to use decorator for that).

It's a custom casting (where we cast a boolean string representation to it's boolean counterpart), and it'll be nice if name somehow suggest that.

I think `strToBool` is probably best

Comment (miguel-a-calles-mba) [reply docs/providers/aws/guide/variables.md]: I made the changes.

File: lib/classes/Variables.js [hunk 3]
```diff
@@ -822,6 +827,17 @@ class Variables {
       });
   }
 
+  getValueStrToBool(variableString) {
+    const groups = variableString.match(this.strToBoolRefSyntax);
+    const param = groups[1];
+    if (param === 'true') {
+      return BbPromise.resolve(true);
```
Comment (medikoo) [inline lib/classes/Variables.js]: Instead of using `BbPromise.resolve`, let's wrap whole body with `BbPromise.try`.

It'll make futer migration to async/await easier, and we don't have to deal with noise of `BbPromise.resolve` and `BbPromise.reject` in code.

---  

Good task for other PR, would be to allow resolver functions to be both `sync` and `async`. As I see it'll be purely about ensuring that non promises are handled without an error in [`promiseTracker.add`](https://github.com/serverless/serverless/blob/46aa304f5df49a651f0899e35534c56a4eeb2cdf/lib/classes/PromiseTracker.js#L39).

Also whether we need a `PromiseTracker` at all is questionable, as I see, same could be acheived with simple meoization that handles promises properly (that could be case for other PR :)

File: lib/classes/Variables.js [hunk 4]
```diff
@@ -822,6 +827,17 @@ class Variables {
       });
   }
 
+  getValueStrToBool(variableString) {
+    const groups = variableString.match(this.strToBoolRefSyntax);
```
Comment (medikoo) [inline lib/classes/Variables.js]: Are we sure that in 100% cases `variableString` will be type of string? What if it comes from a source that returns a number? 

@pmuens you have more experience with that. What do you think?

File: lib/classes/Variables.js [hunk 5]
```diff
@@ -822,6 +827,17 @@ class Variables {
...
+    const param = groups[1];
+    if (param === 'true') {
+      return BbPromise.resolve(true);
+    } else if (param === 'false') {
+      return BbPromise.resolve(false);
+    }
+    return BbPromise.resolve(true);
```
Comment (medikoo) [inline lib/classes/Variables.js]: Here, I think we should follow with one of below options:
- Throw exception (as provided string didn't reflect boolean string) - _that might to restrictive_
- Return `null` (as no matching boolean counterpart was found)
- Apply language boolean coercion (so e.g. empty strings, or empty non string values (e.g. `null`) are resolved as `false`)

File: lib/classes/Variables.js (hunk 4 above)
Comment (pmuens) [reply lib/classes/Variables.js]: Yes, that's a good point. Right now a `variableString` looks something like this: `cf:test-stack.testOutput` or `ssm:/test/path/to/param`, so it seems like it's always a string (at least after looking at the unit tests).

However it could be the case that this changes in the future (not entirely sure why, but if we bloat the Variables implementation we might run into such situation).

File: lib/classes/Variables.js (hunk 5 above)
Comment (miguel-a-calles-mba) [reply lib/classes/Variables.js]: I made updates. Let me know what you think.

File: lib/classes/Variables.js [hunk 6]
```diff
@@ -822,6 +827,22 @@ class Variables {
       });
   }
 
+  getValueStrToBool(variableString) {
+    return BbPromise.try(() => {
+      const groups = variableString.match(this.strToBoolRefSyntax);
```
Comment (medikoo) [inline lib/classes/Variables.js]: I believe we need to take into account that `variableString` may not be string (e.g. internally in other places we already do that: https://github.com/serverless/serverless/blob/56f3c56388faecadc81cf5572e21f98b8b8ae25c/lib/classes/Variables.js#L534 )

And in case it's not a string I would just do `return Boolean(variableString)`.

e.g. I'm pretty sure`${strToBool(strToBool('true'))}` will crash badly if we won't do that (it'll be nice to add test for this case)

File: lib/classes/Variables.js [hunk 7]
```diff
@@ -822,6 +827,22 @@ class Variables {
...
+      if (groups.length < 2) {
+        throw new this.serverless.classes.Error(`${variableString} cannot have an empty string`);
+      }
+      const param = groups[1].trim();
+      if (param === 'false') {
+        return false;
+      } else if (!param || param === '' || param === 'null') {
```
Comment (medikoo) [inline lib/classes/Variables.js]: I wouldn't add special treatment for `'null'`, and other cases seems not possible (so I would get rid of this condition)

File: lib/classes/Variables.js [hunk 8]
```diff
@@ -822,6 +827,22 @@ class Variables {
...
+        throw new this.serverless.classes.Error(`${variableString} resolved to null`);
+      }
+      return true;
```
Comment (medikoo) [inline lib/classes/Variables.js]: I'd add comment, that it's only non-empty string that can end here

File: lib/classes/Variables.js [hunk 9]
```diff
@@ -822,6 +827,22 @@ class Variables {
       });
   }
 
+  getValueStrToBool(variableString) {
+    return BbPromise.try(() => {
+      const groups = variableString.match(this.strToBoolRefSyntax);
+      if (groups.length < 2) {
```
[the remark this answers is not in the document]
Comment (miguel-a-calles-mba) [reply lib/classes/Variables.js]: Makes sense.

File: lib/classes/Variables.js (hunk 6 above)
Comment (miguel-a-calles-mba) [reply lib/classes/Variables.js]: I found using `.match(this.variableSyntax)` does not work as expected.

Comment (miguel-a-calles-mba) [reply lib/classes/Variables.js]: This fails to match `variableString.match(this.variableSyntax)`.
This fails to match `param.match(this.variableSyntax)`.

Comment (medikoo) [reply lib/classes/Variables.js]: My point was not about `value.match(this.variableSyntax)` but about `_.isString(value)`

File: lib/classes/Variables.js (hunk 8 above)
Comment (miguel-a-calles-mba) [reply lib/classes/Variables.js]: Makes sense.

File: lib/classes/Variables.js (hunk 7 above)
Comment (miguel-a-calles-mba) [reply lib/classes/Variables.js]: Makes sense.

File: lib/classes/Variables.js [hunk 10]
```diff
@@ -822,6 +827,24 @@ class Variables {
...
   }
 
+  getValueStrToBool(variableString) {
+    return BbPromise.try(() => {
+      if (_.isString(variableString)) {
+        const groups = variableString.match(this.strToBoolRefSyntax);
+        const param = groups[1].trim();
+          if (param === 'false') {
+            return false;
+          } else if (!param || param === '') {
```
Comment (medikoo) [inline lib/classes/Variables.js]: I think this is dead condition, as by regex param can't turn `''`, am I missing something?

File: lib/classes/Variables.test.js [hunk 11]
```diff
@@ -2378,6 +2378,42 @@ module.exports = {
...
+    });
+    it('should return true (boolean) for a truthy string', () => {
+      return serverless.variables.getValueStrToBool('strToBool(anything)').should.become(true);
+    });
+    it('should throw error for a null-like string', () => {
+      return serverless.variables.getValueStrToBool('strToBool(null)').should.become(true);
+    });
+    it('should throw error for string with spaces only', () => {
+      return serverless.variables
+        .getValueStrToBool('strToBool( )')
```
Comment (medikoo) [inline lib/classes/Variables.test.js]: This if I see correctly will never end in resolver as it doesn't match regex.

Either's there's no point in testing it, or regex should be updated to accept empty strings (or string padding)

File: lib/classes/Variables.js (hunk 6 above)
Comment (miguel-a-calles-mba) [reply lib/classes/Variables.js]: Ahh. Thanks for clarifying.

File: lib/classes/Variables.js (hunk 10 above)
Comment (miguel-a-calles-mba) [reply lib/classes/Variables.js]: Good point.

File: lib/classes/Variables.test.js (hunk 11 above)
Comment (miguel-a-calles-mba) [reply lib/classes/Variables.test.js]: Good point.

File: lib/classes/Variables.js [hunk 12]
```diff
@@ -822,6 +827,24 @@ class Variables {
...
   }
 
+  getValueStrToBool(variableString) {
+    return BbPromise.try(() => {
+      if (_.isString(variableString)) {
+        const groups = variableString.match(this.strToBoolRefSyntax);
+        const param = groups[1].trim();
+        if (param === 'false') {
+          return false;
+        } else if (!param) {
```
Comment (medikoo) [inline lib/classes/Variables.js]: I think it's a dead condition, as regex doesn't allow empty (or just whitespace content)

So either we should remove that condition, or relax regex to allow empty (or just whitespace) strings

Comment (miguel-a-calles-mba) [reply lib/classes/Variables.js]: Sounds good. I removed the condition.

**Comment by miguel-a-calles-mba** [comment]
I'm good with adding a check to convert '0' to false.

**Comment by pmuens** [comment]
> Thanks @miguel-a-calles-mba (and sorry that this PR is dragging so much, we just to be sure it's done right - once it's merged and published, it's hard to follow with breaking changes)

Echoing that. Thanks @miguel-a-calles-mba for your work on this! We're pretty close

> Idea is to allow only following strings : 'true' and 'false' (still I think it's fine to also allow '1' and '0', @pmuens?), and crash with meaningful error on others.

Yes, I think that's fine. So it should be `true`, `false` and `0` and `0`. I wouldn't support something like `null` or `undefined` since those values are programming language dependent.

> with message like: Unexpected `strToBool` input, expected either 'true' or 'false'

In that case we might want to update the error message 🔝 as well (to include `0` and `1`).

**Comment by miguel-a-calles-mba** [comment]
No problem. It's been a great learning experience.
